utils/frame/try-runtime/cli/src/commands/follow_chain.rs

```diff
@@ -42,51 +46,177 @@ pub struct FollowChainCmd {
 	uri: String,
 }
 
+/// Start listening for with `SUB` at `url`.
+///
+/// Returns a pair `(client, subscription)` - `subscription` alone will be useless, because it
+/// relies on the related alive `client`.
+async fn start_subscribing<Header: DeserializeOwned>(url: &str) -> (Client, Subscription<Header>) {
+	let client = WsClientBuilder::default()
+		.connection_timeout(std::time::Duration::new(20, 0))
+		.max_notifs_per_subscription(1024)
+		.max_request_body_size(u32::MAX)
+		.build(url)
+		.await
+		.unwrap();
+
+	log::info!(target: LOG_TARGET, "subscribing to {:?} / {:?}", SUB, UN_SUB);
+
+	let sub = client.subscribe(SUB, None, UN_SUB).await.unwrap();
+	(client, sub)
+}
+
+/// Abstraction over RPC calling for headers.
+#[async_trait]
+trait HeaderProvider<Block: BlockT>
```

Comment:
what is your rationale for creating these two traits and structs? I think `struct FinalizedHeaders` is perfectly fine, but the rest of it rings a bit of over-engineering warning in my head. 

For example, the entire boilerplate around `HeaderProvider` is simply a call into `rpc_api::get_header::<Block, _>(&self.uri, hash)` which is in itself a fine functional abstraction to use.

Reply:
the only reason is testability - since this fetching is not so trivial, I wanted to attach some unit tests; these wouldn't be possible without mocking ws subscription and RPC server